### ISSUE TITLE
Use `angular.equals()` to check if one way bindings changed

### ISSUE THREAD

**Issue by adambuczynski** [angular/angular.js#14827]
**What is the current behavior?**
When a one-way bound array reference changes, even if the `currentValue` and `previousValue` pass the equality check of `angular.equals()`, it is still registered as a change.

This is causing my `$onChanges` even listener to fire even though the current and previous values are of the same type and of the exact same contents/properties.

I am not sure if this is intentional behaviour or a bug.

**What is the expected behavior?**
That this kind of change would not trigger the `$onChanges` event for that particular property, because, other than the reference to another object, nothing has in fact changed (array content wise).

**What is the motivation / use case for changing the behavior?**
My current work around is that I have to manually check the current/previous value with `angular.equals()` to make sure I don't mark the form dirty for no reason (it is used in a form control directive). This feels a bit redundant.

**Which versions of Angular**
Angular 1.5.6

**Comment by gkalpak** [comment]
This is working as intended. For more context see #14378.

tl;dr
Deep-watching would be too expensive and inconsistent with Angular 2.

**Comment by gkalpak** [comment]
BTW, the possibility to add supporting for `$watchCollection` to one-way bindings is discussed in #14039 (as you already know :wink:). It's not equivalent to deep-watching, but it is deeper than watching by reference.

**Comment by adambuczynski** [comment]
Yeah, I figured this would be the case. Ok, I'll use `angular.equals()` as a workaround for now.
